### PR TITLE
Update SmartThings translators to supprot lightbulbs and thermostat in new resource model.

### PR DESCRIPTION
**Pull request by m-chau**
•	Updated SmartThings lightbulb and thermostat translators to match the new resource model.
•	Updated SmartThings hub to support lightbulb and thermostat translators in new resource model.
•	Update the SmartApp source code to support thermostat and subscribe/unsubscribe to event of a specific device.
•	Move things in new resource model from “opent2t.p.*” folders to “org.opent2t.sample.*.superpopular” folders (and updated the schema names too).
•	Enabled color and color temperature change.

### CODE REVIEW

**Comment by arjun-msft** [comment]
>             /**

don't we need to uncomment these?

---
Refers to: org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js:132 in 73a2d08. [](commit_id = 73a2d08ae45596e8479bc283eb8751b688e297a4, deletion_comment = False)

File: org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js [hunk 1]
```diff
@@ -161,6 +165,24 @@ class Translator {
         });
     }
 
+    _subscribe(deviceId) {
```
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js]: This doesn't appear to abide by the schema: _subscribe(deviceType, deviceId, callbackUrl, verificationRequest)

File: org.opent2t.sample.lamp.superpopular/com.smartthings.lightbulb/js/thingTranslator.js [hunk 2]
```diff
@@ -13,101 +13,202 @@ function validateArgumentType(arg, argName, expectedType) {
...
+// This translator class implements the 'opent2t.d.light' interface.
 class Translator {
 
     constructor(deviceInfo) {
         console.log('SmartThings Lightbulb initializing...');
 
         validateArgumentType(deviceInfo, "deviceInfo", "object");
         
-        deviceId = deviceInfo.deviceInfo.id;
+        deviceId = deviceId = deviceInfo.deviceInfo.opent2t.controlId;
```
Comment (arjun-msft) [inline org.opent2t.sample.lamp.superpopular/com.smartthings.lightbulb/js/thingTranslator.js]: >= [](start = 17, length = 1)

typo, don't need double deviceId =

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+    getDevicesFanMode(di) {
+        return getDeviceResource(this, di, 'fanMode');
+    }
+
+    postDevicesFanMode(di, payload) {
+        return postDeviceResource(di, 'fanMode', payload);
+    }
+
+    postSubscribe(callbackUrl, verificationRequest) {
```
Comment (arjun-msft) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: >callbackUrl [](start = 18, length = 11)

how come the callbackUrl isn't used?

**Comment by arjun-msft** [comment]
:clock1:

File: org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js (hunk 1 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js]: In the current design, the callback URL are set in the SmartApp as last time when I had the conversation with Avneet, he mentioned that we will not be dynamically changing the callback URL at the current stage. However, the current SmartApp design does provide a option to update the callback URL through a REST API call if needed.

File: org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/tests/test.js [hunk 4]
```diff
@@ -24,10 +24,10 @@ function getBinarySwitch(devices) {
 // setup the translator before all the tests run
 test.before(async () => {
     var hubTranslator = await OpenT2T.createTranslatorAsync(hubPath, 'thingTranslator', config);
-    var hubInfo = await OpenT2T.invokeMethodAsync(hubTranslator, 'org.opent2t.sample.hub.superpopular', 'get', [false]);
```
Comment (m-chau) [inline org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/tests/test.js]: Should I change the test files occurring to Chuck's PR right now?

**Comment by m-chau** [comment]
>             /**

Fixed.

---
In reply to: [264081136](https://github.com/openT2T/translators/pull/134#issuecomment-264081136) [](ancestors = 264081136)

---
Refers to: org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js:132 in 73a2d08. [](commit_id = 73a2d08ae45596e8479bc283eb8751b688e297a4, deletion_comment = False)

File: org.opent2t.sample.lamp.superpopular/com.smartthings.lightbulb/js/thingTranslator.js (hunk 2 above)
Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.smartthings.lightbulb/js/thingTranslator.js]: Fixed.

---
In reply to: [90384829](https://github.com/openT2T/translators/pull/134#discussion_r90384829) [](ancestors = 90384829)

File: org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js (hunk 1 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js]: In the current design, the callback URL are set in the SmartApp as last time when I had the conversation with Avneet, he mentioned that we will not be dynamically changing the callback URL at the current stage. However, the current SmartApp design does provide a option to update the callback URL through a REST API call if needed.

---
In reply to: [90384719](https://github.com/openT2T/translators/pull/134#discussion_r90384719) [](ancestors = 90384719)

File: org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/tests/test.js (hunk 4 above)
Comment (m-chau) [inline org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/tests/test.js]: Should I change the test files occurring to Chuck's PR right now?

---
In reply to: [90384586](https://github.com/openT2T/translators/pull/134#discussion_r90384586) [](ancestors = 90384586)

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 3 above)
Comment (m-chau) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: In the current design, the callback URL is set in the SmartApp.

---
In reply to: [90385111](https://github.com/openT2T/translators/pull/134#discussion_r90385111) [](ancestors = 90385111)

File: org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js (hunk 1 above)
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/thingTranslator.js]: [https://github.com/openT2T/translators/issues/137](https://github.com/openT2T/translators/issues/137)


---
In reply to: [90439673](https://github.com/openT2T/translators/pull/134#discussion_r90439673) [](ancestors = 90439673,90384719)

**Comment by arjun-msft** [comment]
Only thing left is to update your PR with the latest changes in the branch

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 5]
```diff
@@ -116,8 +116,8 @@ function providerSchemaToPlatformSchema(providerSchema, expand) {
             controlId: controlId
         },
         pi: providerSchema['id'],
-        mnmn: providerSchema['manufacturer'],
-        mnmo: providerSchema['model'],
+        mnmn: providerSchema['manufacturer'] === null ? "Unkown" : providerSchema['manufacturer'],
```
Comment (shalinijoshi19) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: >Unkown [](start = 57, length = 6)

typo - 'Unknown'

Comment (shalinijoshi19) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: > === null ? "Unkown" [](start = 44, length = 20)

Can we not use undefined here?

**Comment by shalinijoshi19** [comment]
:clock1:

File: org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js [hunk 6]
```diff
@@ -34,8 +34,8 @@ function providerSchemaToPlatformSchema(providerSchema, expand) {
             controlId: controlId
         },
         pi: providerSchema['id'],
-        mnmn: providerSchema['manufacturer'],
-        mnmo: providerSchema['model'],
+        mnmn: providerSchema['manufacturer'] === null ? "Unkown" : providerSchema['manufacturer'],
```
Comment (arjun-msft) [inline org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js]: spelling: Unknown

File: org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js [hunk 7]
```diff
@@ -34,8 +34,8 @@ function providerSchemaToPlatformSchema(providerSchema, expand) {
...
+        mnmo: providerSchema['model'] === null ? "Unkown" : providerSchema['model'],
```
Comment (arjun-msft) [inline org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js]: spelling: Unknown

File: org.opent2t.sample.lamp.superpopular/com.smartthings.lightbulb/js/thingTranslator.js [hunk 8]
```diff
@@ -194,8 +194,8 @@ function providerSchemaToPlatformSchema(providerSchema, expand) {
             controlId: controlId
         },
         pi: providerSchema['id'],
-        mnmn: providerSchema['manufacturer'],
-        mnmo: providerSchema['model'],
+        mnmn: providerSchema['manufacturer'] === null ? "Unkown" : providerSchema['manufacturer'],
+        mnmo: providerSchema['model'] === null ? "Unkown" : providerSchema['model'],
```
Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.smartthings.lightbulb/js/thingTranslator.js]: Changed to return 'Undefined'.

---
In reply to: [92027584](https://github.com/openT2T/translators/pull/134#discussion_r92027584) [](ancestors = 92027584)

File: org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js (hunk 6 above)
Comment (m-chau) [inline org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js]: Changed to return 'Undefined'.

---
In reply to: [92070701](https://github.com/openT2T/translators/pull/134#discussion_r92070701) [](ancestors = 92070701)

File: org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js (hunk 7 above)
Comment (m-chau) [inline org.opent2t.sample.binaryswitch.superpopular/com.smartthings.binaryswitch/js/thingTranslator.js]: Changed to return 'Undefined'.

---
In reply to: [92070717](https://github.com/openT2T/translators/pull/134#discussion_r92070717) [](ancestors = 92070717)

**Comment by m-chau** [comment]
Done.

---
In reply to: [264533380](https://github.com/openT2T/translators/pull/134#issuecomment-264533380) [](ancestors = 264533380)

**Pull request merged**